### PR TITLE
refactor: Auth Actions (action classes are executed in the order set in `Config/Auth::$actions`)

### PR DESCRIPTION
**Pull request by kenjis**
~~Need to rebase after merging #391~~

Change the behavior of auth actions based on the following assumptions:
- All actions must be processed one at a time in sequence
- Action classes are executed in the order set in `Config/Auth::$actions`
- All actions have identities

### CODE REVIEW

**Comment by MGatner** [comment]
I'll need to come back to these two another day.

**Comment by MGatner** [comment]
Awaiting rebase.

**Review by MGatner — APPROVED** [review]
Looks good to me. I'm still not actually using Shield in any projects so let's be sure to get another review from someone. @datamweb and/or @lonnieezell?

**Review by datamweb — APPROVED** [review]
@kenjis code works fine, and I didn't see any problems.
**Thanks**
During the review of this PR, I noticed an issue, which I thought was better to raise here.
1. Register a new user.By default, every user login the system after registration.
2 . Check the table with command `php spark db:table test__auth_logins`.
Why is the user's login not recorded in the table?!!
It seems that we need PR to cover this issue.

**Comment by kenjis** [comment]
> Why is the user's login not recorded in the table?!!

It is intentional.
Because it is not a login to record. It is a registration.
A login to record is a user action to login.
I think that a user does not think they are logged in when they register.

The current specification is stated in #131

**Comment by datamweb** [comment]
> It is intentional. 

Thanks Kenjis, I will check #131.

**Comment by kenjis** [comment]
Updated the docs.

**Comment by kenjis** [comment]
I refactored action classes to be more extensible. 

Now you can probably extend the action class to use time limited multiple codes:
https://github.com/codeigniter4/shield/discussions/343#discussioncomment-3416568

**Review by MGatner — APPROVED** [review]
Looks good! I'm glad to see some of the ambiguity cleared up. I can't keep track of what has been released... Do we need to start a changelog for some of these behavior changes?

**Review by datamweb — CHANGES_REQUESTED** [review]
I think the following change will help the readability of the documents(`auth_actions.md`)
1. **Email-based Account Activation** 
2. **Email-based Two Factor Authentication** 

I prefer to check this PR locally more carefully so I need more time.
If you are not in a hurry to merge, wait a little.

File: tests/Unit/EmailActivatorTest.php [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+/**
+ * @internal
+ */
+final class EmailActivatorTest extends TestCase
+{
+    public function testShowCannotGetPendingLoginUser(): void
+    {
+        $this->expectException(RuntimeException::class);
+        $this->expectExceptionMessage('Cannot get the pending login User');
```
Comment (datamweb) [inline tests/Unit/EmailActivatorTest.php:20]:
```suggestion
        $this->expectExceptionMessage('Cannot get the pending login User.');
```

File: tests/Unit/EmailActivatorTest.php [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+        $activator = new EmailActivator();
+
+        $activator->show();
+    }
+
+    public function testVerifyCannotGetPendingLoginUser(): void
+    {
+        $this->expectException(RuntimeException::class);
+        $this->expectExceptionMessage('Cannot get the pending login User');
```
Comment (datamweb) [inline tests/Unit/EmailActivatorTest.php:30]:
```suggestion
        $this->expectExceptionMessage('Cannot get the pending login User.');
```

**Review by datamweb — CHANGES_REQUESTED** [review]

File: tests/Unit/EmailActivatorTest.php (hunk 1 above)
Comment (kenjis) [reply tests/Unit/EmailActivatorTest.php]: Done.

**Comment by kenjis** [comment]
@datamweb 
> I prefer to check this PR locally more carefully so I need more time.

Okay. I don't hurry.

**Review by datamweb — APPROVED** [review]
Well! 
After hours of playing with the code, I found no problem.

**Comment by kenjis** [comment]
@datamweb Thank you for your testing!

**Pull request merged**